docs/devel/consumers-claims.md

```diff
@@ -0,0 +1,297 @@
+`VolumeManager`s running, each of which handles some set of claim classes.
+When a `PersistentVolumeClaim` is created that asks for "Gold" storage, the
+`VolumeManager` which understands the "Gold" class will make the binding.
+
+`VolumeManager`s can be simple `Pod`s running in the cluster, pulling
+cofiguration from hardcoded values, flags, config objects, or whatever
+mechanism makes sense for them.  We expect config objects to be a perfect fit
+for this use case.
+
+This very loose coupling means that an admin can define classes however they
+like.  For example, "Gold" might mean GCE PersistentDisk of type SSD, "Silver"
+might mean GCE PersistentDisk of type Disk, and "Bronze" might mean NFS.
+Configuring the same `VolumeManager` to run for both "Gold" and "Silver" with
+just a change in params is trivial, as would be adding a totally bespoke
+`VolumeManager`.
+
+Note that classes are just identifiers - there are no special names or
+meanings.  One of the `VolumeManager`s must be the "default" for claims that do
+not specify a class.
+
+### Config
+
+We expect that `VolumeManager`s will need some amount of configuration, and it
+will be convenient if they have similar parameters, but this should be by
+convention, not by diktat.  Example parameters:
+   * Minimum free instances - how many ready-to-use volumes should be maintained
+   * Maximum free instances - limit of number of ready-to-use volumes
+   * Size distribution - desired histogram of sizes of ready-to-use volumes
+
+In addition to these, there will be volume-type specific parameters, such as
+the GCE SSD vs Disk param described above.
+
+### Scrubbing
+
+Volumes need to be scrubbed before they can be reused, and some storage devices
+may even need to be scrubbed before they are decommissioned.  This is the
+responsibility of the `VolumeManager` as it drives the lifecycle.
+
+## Networking
+
+TODO: flip Ingress and LB ?
+
+Network ingress in Kubernetes will be represented by `LoadBalancer`s.  A user
+can create an `IngressClaim` and reference that claim from an `Ingress` object.
+A `LoadBalancerManager` is responsible for binding `IngressClaim`s to
+`LoadBalancers`s.  If the `IngressClaim` can not be satisified by any existing
+`LoadBalancer`, the `LoadBalancerManager` may provision a new one.
+
+```
+                       +--------------+
+        +--create----> | LoadBalancer |
+        |              +--------------+
+   +----+----+               ^
+   |  LBMgr  |---bind------> :
+   +----+----+               v
+        |              +---------------+
+        +---watch----> | IngressClaim  | <--+
+                       +---------------+    |
+                                 ^          |
+                                 |          +-------- user-created
+                                 |          |
+                           +----------+     |
+                           | Ingress  | <---+
+                           +----------+
+```
+
+### IngressClaim
```

Comment:
It does not - admins create ingress rules for their services.

On Sep 17, 2015, at 1:26 AM, David Oppenheimer <notifications@github.com>
wrote:

In docs/devel/consumers-claims.md
<https://github.com/kubernetes/kubernetes/pull/14095#discussion_r39712333>:

> +        |              +--------------+
> +   +----+----+               ^
> +   |  LBMgr  |---bind------> :
> +   +----+----+               v
> +        |              +---------------+
> +        +---watch----> | IngressClaim  | <--+
> +                       +---------------+    |
> +                                 ^          |
> +                                 |          +-------- user-created
> +                                 |          |
> +                           +----------+     |
> +                           | Ingress  | <---+
> +                           +----------+
> +```
> +
> +### IngressClaim

Or to be more precise: does a Pod reference an Ingress by name?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/14095/files#r39712333>.